### DISCUSSION TITLE
How to implement XAML Islands from WinRT bindings?

### DISCUSSION THREAD

**Discussion by ElSaico** [microsoft/WindowsAppSDK#4588] [Q&A]
Greetings!

I'm working on a massive UI refactoring on a Python application, and one issue that needs tackling is the Windows title bar.
For the app's dark theme, calling the DWM Win32 API would suffice, but it also has a _transparent_ theme - the current implementation disables the window decoration and renders its own title bar from scratch, which requires manually handling basic stuff like dragging and caption buttons, and demands gnarly workarounds for things like minimizing to tray.

[The Windows App SDK should be the way to go for this](https://learn.microsoft.com/en-us/windows/apps/develop/title-bar) - meaning [PyWinRT](https://pywinrt.readthedocs.io/en/latest/) to the rescue! - but as mentioned there it requires making a custom XAML component to have transparency... or changing colors at all on Windows 10, for that matter. XAML Islands are an obvious solution, yet I faced the same issues as #2233 with the available resources mixing up UWP and WASDK and centering on plain C++ Win32 apps.

Given the circumstances, is it viable to create and attach an island through bindings such as PyWinRT?

**Comment by DarranRowe** [comment]
> The Windows App SDK should be the way to go for this - meaning PyWinRT to the rescue! - but as mentioned there it requires making a custom XAML component to have transparency... or changing colors at all on Windows 10, for that matter.

First of all, this isn't true. Simple customisation isn't supported on Windows 10, but there are multiple ways to do this without using Xaml. The big thing to remember is that the custom titlebar functionality in AppWindow is just an implementation of the DWM method. This means that any method of drawing to the top of the window will work.
I have personally done this using Windows.UI.Composition and Direct2D, but using layered windows and GDI should work if you use the alpha channel properly.

> Given the circumstances, is it viable to create and attach an island through bindings such as PyWinRT?

There is a reason why the example is set up to use the Xaml compiler. It is possible to implement things without using the Xaml compiler, but it isn't the nicest thing to do. The ability to create WinRT components is required for full Xaml functionality.
It is possible to create Xaml Island content missing some things, but the more you are missing, the more functionality is lost. To give one basic example of this, to create an instance of XamlControlsResources and apply the styles to Xaml controls requires an active instance of the Xaml application that you can query for IXamlMetadataProvider. This IXamlMetadataProvider interface must then provide the type information from XamlControlsXamlMetadataProvider.
Unfortunately, while I am able to provide information on the requirements, I am unable to say if it is viable to use islands in Python or to what extent. Hopefully someone with more experience in this matter is able to guide you, but you should be ready to implement WinRT components if you follow this path.

**Reply by ElSaico** [reply]
Thank you very much for your answer! The composition route seems promising - is this such an example?
https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/using-the-visual-layer-with-win32

**Reply by DarranRowe** [reply]
Yes, that is the sample for using this API.
